fdb-record-layer-core/src/main/java/com/apple/foundationdb/record/provider/foundationdb/FDBRecordStore.java

```diff
@@ -476,6 +498,47 @@ protected <M extends Message> CompletableFuture<FDBStoredRecord<M>> saveTypedRec
         return context.instrument(FDBStoreTimer.Events.SAVE_RECORD, result);
     }
 
+    @Nonnull
+    @API(API.Status.INTERNAL)
+    public <M extends Message> CompletableFuture<FDBStoredRecord<M>> dryRunSaveRecordAsync(@Nonnull M rec,
+                                                                                    @Nonnull RecordExistenceCheck existenceCheck) {
+        final RecordMetaData metaData = metaDataProvider.getRecordMetaData();
+        final Descriptors.Descriptor recordDescriptor = rec.getDescriptorForType();
+        final RecordType recordType = metaData.getRecordTypeForDescriptor(recordDescriptor);
+        final KeyExpression primaryKeyExpression = recordType.getPrimaryKey();
+
+        final FDBStoredRecordBuilder<M> recordBuilder = FDBStoredRecord.newBuilder(rec).setRecordType(recordType);
+        final FDBRecordVersion recordVersion = recordVersionForSave(metaData, null, VersionstampSaveBehavior.DEFAULT);
+        recordBuilder.setVersion(recordVersion);
+        final Tuple primaryKey = primaryKeyExpression.evaluateSingleton(recordBuilder).toTuple();
+        recordBuilder.setPrimaryKey(primaryKey);
+
+        final CompletableFuture<FDBStoredRecord<M>> result = loadExistingRecord(serializer, primaryKey).thenCompose(oldRecord -> {
+            if (oldRecord == null) {
+                if (existenceCheck.errorIfNotExists()) {
+                    throw new RecordDoesNotExistException("record does not exist",
+                            LogMessageKeys.PRIMARY_KEY, primaryKey);
+                }
+            } else {
+                if (existenceCheck.errorIfExists()) {
+                    throw new RecordAlreadyExistsException("record already exists",
+                            LogMessageKeys.PRIMARY_KEY, primaryKey);
+                }
+                if (existenceCheck.errorIfTypeChanged() && oldRecord.getRecordType() != recordType) {
+                    throw new RecordTypeChangedException("record type changed",
+                            LogMessageKeys.PRIMARY_KEY, primaryKey,
+                            LogMessageKeys.ACTUAL_TYPE, oldRecord.getRecordType().getName(),
+                            LogMessageKeys.EXPECTED_TYPE, recordType.getName());
+                }
+            }
+            // (TODO): recordBuilder.setSizeInfo(sizeInfo);
+            // (TODO): update indexState in recordStoreState
+            return CompletableFuture.completedFuture(recordBuilder.build());
+        });
+        // (TODO): timer should be changed here
+        return context.instrument(FDBStoreTimer.Events.SAVE_RECORD, result);
+    }
```

Comment:
This is a verbatim copy of `saveTypedRecord` without this line:

```java
final FDBStoredRecord<M> newRecord = serializeAndSaveRecord(typedSerializer, recordBuilder, metaData, oldRecord);
```

I think it makes sense to think about some refactoring here to reduce code duplication, e.g. by means of introducing a boolean flag in `saveTypedRecord`, or some sort of `read-only` attribute defined in `FDBRecordStore` that is checked in `saveTypedRecord` s.t. we conditionally call `serializeAndSaveRecord`.

Regardless, I think this should be well documented.